libraries/adaptive-expressions/src/functionUtils.ts

```diff
@@ -0,0 +1,1033 @@
+     */
+    public static parseStringOrNull(input: string | undefined): string {
+        if (typeof input === 'string') {
+            return input;
+        } else {
+            return '';
+        }
+    }
+
+    /**
+     * Try to accumulate the path from an Accessor or Element, from right to left
+     * return the accumulated path and the expression left unable to accumulate
+     * @param expression
+     * @param state scope
+     * @param options Options used in evaluation
+     */
+    public static tryAccumulatePath(expression: any, state: MemoryInterface, options: Options): {path: string; left: any; error: string} {
+        let path = '';
+        let left = expression;
+        while (left !== undefined) {
+            if (left.type === ExpressionType.Accessor) {
+                path = (left.children[0] as Constant).value + '.' + path;
+                left = left.children.length === 2 ? left.children[1] : undefined;
+            } else if (left.type === ExpressionType.Element) {
+                let value: any;
+                let error: string;
+                ({value, error} = left.children[1].tryEvaluate(state, options));
+
+                if (error !== undefined) {
+                    return {path: undefined, left: undefined, error};
+                }
+
+                if (FunctionUtils.isNumber(parseInt(value))) {
+                    path = `[${value}].${path}`;
+                } else if (typeof value === 'string') {
+                    path = `['${value}'].${path}`;
+                } else {
+                    return {path: undefined, left: undefined, error: `${left.children[1].toString()} doesn't return an int or string`};
+                }
+
+                left = left.children[0];
+            } else {
+                break;
+            }
+        }
+
+        // make sure we generated a valid path
+        path = path.replace(/(\.*$)/g, '').replace(/(\.\[)/g, '[');
+        if (path === '') {
+            path = undefined;
+        }
+
+        return {path, left, error: undefined};
+    }
+
+    /**
+     * Get the value of a path from a memory
+     * @param state memory.
+     * @param path path string.
+     * @param options options.
+     */
+    public static wrapGetValue(state: MemoryInterface, path: string, options: Options): any {
+        let result = state.getValue(path);
+        if (result !== undefined && result !== null) {
+            return result;
+        }
+
+        if (options.nullSubstitution !== undefined) {
+            return options.nullSubstitution(path);
+        }
+
+        return undefined;
+    }
+
+    /**
+     * Evaluator for foreach and select function
+     */
+    public static foreach(expression: any, state: MemoryInterface, options: Options): {value: any; error: string} {
+        let result: any[];
+        let error: string;
+        let instance: any;
+
+        ({value: instance, error} = expression.children[0].tryEvaluate(state, options));
```

Comment:
>({value: instance, error} = expression.children[0].tryEvaluate(state, options)); [](start = 8, length = 80)

Extra parens

Reply:
It is a typescript feature that can assign the result to variables.

Reply:
Wouldn't it be clearer as:
let {value, error} = expression.children...

---
In reply to: [454837340](https://github.com/microsoft/botbuilder-js/pull/2502#discussion_r454837340) [](ancestors = 454837340)

Reply:
It's ok. But sometimes we just want to use an alias to describe the variable, value -> instance.
Both way is fine. 
```
let {value, error} = expression.children...
const instance = value;
```
or 
```
let instance;
let error;
({value: instance, error} = expr...
```